### PR TITLE
STAC-15323 configmap data limit

### PR DESCRIPTION
**Pull request by Q69K**
Cut configmap data to a configurable limit

### CODE REVIEW

File: pkg/collector/corechecks/cluster/kubeapi/kubernetes_topology_config.go [hunk 1]
```diff
@@ -21,20 +21,29 @@ const (
 	OpenShift = "openshift"
 )
 
+const (
+	DefaultConfigMapDataLimit = 100 * 1024
```
Comment (julianosk) [inline pkg/collector/corechecks/cluster/kubeapi/kubernetes_topology_config.go:25]: This const could go with the ones above

File: pkg/collector/corechecks/cluster/topologycollectors/configmap_collector.go [hunk 2]
```diff
@@ -62,9 +68,62 @@ func (cmc *ConfigMapCollector) configMapToStackStateComponent(configMap v1.Confi
...
+}
+
+// cutData tries to reduce `data` size
+// it replaces values within the map completely or partially with `[dropped N, hash...]` string
+// it's intended to strip some big files out of configmap,
+// and on scale of hundred of KiB it's precise enough: it doesn't take into account replacement string
+// Algorithm:
+// 1. remove values that are on their own are bigger than limit
+// 2. proportionally cut the rest of the values
+func cutData(data map[string]string, maxSize int) map[string]string {
```
Comment (julianosk) [inline pkg/collector/corechecks/cluster/topologycollectors/configmap_collector.go:98]: I wonder if this complexity is really needed. Let's discuss with Mark.

**Review by julianosk — COMMENTED** [review]
Let's discuss with Mark.

File: pkg/collector/corechecks/cluster/kubeapi/kubernetes_topology_config.go [hunk 3]
```diff
@@ -21,20 +21,29 @@ const (
...
+	CheckID              check.ID
+	Instance             topology.Instance
 }
 
 func (c *TopologyConfig) parse(data []byte) error {
 	// default values
 	c.ClusterName = config.Datadog.GetString("cluster_name")
 	c.CollectTopology = config.Datadog.GetBool("collect_kubernetes_topology")
 	c.CollectTimeout = config.Datadog.GetInt("collect_kubernetes_timeout")
+	c.ConfigMapMaxDataSize = config.Datadog.GetSizeInBytes("configmap_max_datasize")
```
Comment (julianosk) [inline pkg/collector/corechecks/cluster/kubeapi/kubernetes_topology_config.go:43]: Add option to helm chart

File: pkg/collector/corechecks/cluster/topologycollectors/configmap_collector.go (hunk 2 above)
Comment (Q69K) [reply pkg/collector/corechecks/cluster/topologycollectors/configmap_collector.go:94]: old implementation indeed has disadvantages (it was unstable - if one size of data in a single key is changed, then all keys were cut differently) - therefore I updated to simpler algorithm we discussed

File: pkg/collector/corechecks/cluster/kubeapi/kubernetes_topology_config.go [hunk 4]
```diff
@@ -19,22 +20,29 @@ const (
 	Kubernetes ClusterType = "kubernetes"
 	// OpenShift cluster type
 	OpenShift = "openshift"
+
```
Comment (julianosk) [inline pkg/collector/corechecks/cluster/kubeapi/kubernetes_topology_config.go:23]: Maybe golint will complain about a comment here

Comment (Q69K) [reply pkg/collector/corechecks/cluster/kubeapi/kubernetes_topology_config.go:23]: I hope it won't, because it's legit grouping

**Review by julianosk — APPROVED** [review]
LGTM

**Review by julianosk — CHANGES_REQUESTED** [review]
Can we log the data size limit configured?
Also, I think it might be good to sort the keys in the data map, sometimes they change order then it looks like it was updated when it wasn't.

**Comment by Q69K** [comment]
@julianosk there is no point in sorting keys, as JSON defines object as unordered collection - so we can't rely further down the pipeline to have the same ordering whatever we do here.

**Review by julianosk — APPROVED** [review]
LGTM

**Pull request merged**